src/main/java/at/ac/tuwien/kr/alpha/grounder/transformation/impl/StratifiedEvaluation.java

```diff
@@ -0,0 +1,521 @@
+			for (Atom a : rule.getBodyAtomsPositive()) {
+				tmpPredicate = a.getPredicate();
+				tmpInstances = this.workingMemory.get(tmpPredicate, true);
+				this.modifiedInLastEvaluationRun.putIfAbsent(tmpPredicate, new LinkedHashSet<>());
+				if (tmpInstances != null) {
+					this.modifiedInLastEvaluationRun.get(tmpPredicate).addAll(tmpInstances.getAllInstances());
+				}
+			}
+		}
+	}
+
+	private void evaluateRule(InternalRule rule) {
+		LOGGER.debug("Evaluating rule {}", rule);
+		List<Substitution> groundSubstitutions = this.groundRule(rule);
+		boolean canRuleFire;
+		for (Substitution subst : groundSubstitutions) {
+			LOGGER.debug("Checking if rule can fire for substitution: {}", subst);
+			canRuleFire = this.canFire(rule, subst);
+			LOGGER.debug("canFire result = {}", canRuleFire);
+			if (canRuleFire) {
+				this.fireRule(rule, subst);
+			}
+		}
+	}
+
+	private List<Substitution> groundRule(InternalRule rule) {
+		LOGGER.debug("Grounding rule {}", rule);
+		RuleGroundingOrders groundingOrders = rule.getGroundingOrders();
+		List<Substitution> groundSubstitutions = new ArrayList<>(); // the actual full ground substitutions for the rule
+		LOGGER.debug("Is fixed rule? {}", rule.getGroundingOrders().fixedInstantiation());
+		if (groundingOrders.fixedInstantiation()) {
+			groundSubstitutions.addAll(this.bindNextAtomInRule(rule, groundingOrders.getFixedGroundingOrder(), 0, new Substitution()));
+		} else {
+			Collection<Literal> startingLiterals = groundingOrders.getStartingLiterals();
+			for (Literal lit : startingLiterals) {
+				groundSubstitutions.addAll(this.calcSubstitutionsForStartingLiteral(rule, lit));
+			}
+		}
+		return groundSubstitutions;
+	}
+
+	private List<Substitution> calcSubstitutionsForStartingLiteral(InternalRule rule, Literal startingLiteral) {
+		List<Substitution> retVal = new ArrayList<>();
+		Substitution partialStartSubstitution = new Substitution();
+		if (startingLiteral instanceof FixedInterpretationLiteral) {
+			// if the starting literal is a builtin or external, we don't have instances in working memory,
+			// but just get fixed substitutions and let bindNextAtomInRule do it's magic ;)
+			FixedInterpretationLiteral fixedInterpretationLiteral = (FixedInterpretationLiteral) startingLiteral;
+			for (Substitution fixedSubstitution : fixedInterpretationLiteral.getSubstitutions(partialStartSubstitution)) {
+				LOGGER.debug("calling bindNextAtom, startingLiteral = {}", startingLiteral);
+				retVal.addAll(this.bindNextAtomInRule(rule, rule.getGroundingOrders().orderStartingFrom(startingLiteral), 0, fixedSubstitution));
+			}
+		} else {
+			Predicate predicate = startingLiteral.getPredicate();
+			if (!this.modifiedInLastEvaluationRun.containsKey(predicate)) {
+				return retVal;
+			}
+			Set<Instance> instances;
+			if (!this.workingMemory.contains(predicate)) {
+				LOGGER.debug("No instances for starting literal " + predicate.toString() + " in working memory!");
```

Comment:
As we discusses, objects should not be `toString()`-ed on the outside of the logger method, as this creates strings even when nothing is logged.
Suggestion here:
`LOGGER.debug("No instances for starting literal over predicate {} in working memory!", predicate);`

Reply:
Fixed (following a larger refactoring after PR #256 got merged)